### PR TITLE
feat(pod-template): add tor service pod template

### LINKED ISSUE #17: [REQUEST] Support specifying various `PodSpec` properties on the OnionService pods
**Is your feature request related to a problem? Please describe.**
I need to be able to control some spec properties on the onion service pods.  My immediate pain is that I want to ensure the service continues to run even when the cluster comes under memory or CPU pressure, which means I need to be able to specify a higher `priorityClassName` for the pods. 

It would also be nice to be able to:
1. add tolerations
2. set resource requests/limits
3. specify affinity rules

Additionally, I currently don't have any specific use-cases in mind, but I could envision other users wanting to set other pod properties (ex: `labels`, `annotations`, `hostNetwork`, `topologySpreadConstraints`, etc).  See https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#PodSpec for a full list of `PodSpec` properties.

**Describe the solution you'd like**
Add a "template" property to the `OnionService` spec:

```
apiVersion: tor.k8s.torproject.org/v1alpha2
kind: OnionService
metadata:
  name: example-onion-service
spec:
  version: 3
  template:
    spec:
      priorityClassName: high-priority
      tolerations: []
      resources: {}
      affinity: {}
  rules: [...]
```

Rather than manually creating this template spec for this project, it may be best to leverage the existing "[PodTemplateSpec](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-template-v1/#PodTemplateSpec)" (although this may introduce complications/confusion if users try to define the `containers` in the spec?).

**Describe alternatives you've considered**
I considered changing the default `priorityClass` to something higher, and setting all less-crucial workloads to a lower class.  This does not work for me, because there are several other 3rd party projects that don't support controlling their workloads' priority, and the OnionService is the only one I would want to be considered a high-priority class.

**Additional context**
I would love to make Onion Services the primary ingress channel into my cluster (potentially even for access to the control-plane), so I am very interested in trying to make it more robust and reliable.

I would be happy to start on a PR to support this, if you are happy with the strategy.

### PR DESCRIPTION
**Pull request by conneryn**
Fixes #17

### CODE REVIEW

**Comment by bugfest** [comment]
Hi @conneryn, thanks for the changes. 

I think we could improve future templating features if we use corev1.PodSpec (`template.spec`) directly instead of copying parts of it. I defined Reources as part of the template (`template.resources`). Do you like this approach?

proposal at branch https://github.com/bugfest/tor-controller/tree/podspectemplate

**Comment by conneryn** [comment]
Great, thanks for the feedback!

I like the idea of pulling the `resources` property out to the top level of the template. I made that change, and also manually applied most of the changes from your commit above (sorry, didn't actually rebase/work from your commit, as I'd already mid-progress on some of the changes and didn't want to deal with resolving the conflicts).

For the `OnionBalancedService`, I added a `balancerTemplate` property and two separate sub-properties (`torResources` and `balancerResources`) for the two containers that get added to the balancer pod (see [onionbalancedservice-resources.yaml](https://github.com/bugfest/tor-controller/pull/18/files#diff-e8d04388bcef38c4de2c75a5634bdf40470ad697eed6668bc7ec804971caafff) for an example).  
NOTE: I'm not a huge fan of the naming here, so I'm **very** open to suggestions/alternatives.

I also updated the template `spec`s to be `corev1.PodSpec`, so a user can set any properties they want.
However, two issues arise from using `corev1.PodSpec` which I had to introduce "hacky"-ish workarounds for:
1. PodSpec has `containers` set as required which causes validation of OnionService objects to fail unless set.  This is solved by adding [hacks/remove-containers-requirement.sh](https://github.com/bugfest/tor-controller/pull/18/files#diff-dd1a3843800bbce36e2eb1dfa417e20ca1fc59f0df0817766ea4a502a821b406), which runs during `make manifests` to remove the requirement from the generated CRDs.  This solution is borrowed from [here](https://github.com/dippynark/cluster-api-provider-kubernetes/blob/731d0dd6e972b325fc1a3a35931cf6724b17d7bf/hack/remove-containers-requirement.sh).
2. PodSpec is large, and adding it to the OnionBalancedService caused the generated CRDs to be very large.  This makes it so `kubectl apply -f --` fails to apply the CRD with the error "metadata.annotations: Too long: must have at most 262144 bytes" (`kubectl apply` attempts to add an annotation `last-applied-configuration`, with the whole contents of the document). Similar discussions can be found [here](https://medium.com/pareture/kubectl-install-crd-failed-annotations-too-long-2ebc91b40c7d), [here](https://github.com/argoproj/argo-cd/issues/820), [here](https://github.com/kubernetes-sigs/kubebuilder/issues/1140), and [here](https://github.com/kubernetes-sigs/kubebuilder/issues/2556).
    **Current Solution**: Don't include descriptions with the CRD (added `crd:maxDescLen=0` to `make manifests`).
    **Alternate Option**: use `kubectl create` or `replace` instead of `kubectl apply` in the `make install` phase.  If any end-users ever try to manually `kubectl apply` the CRDS, they would still run into this issue.

**Comment by bugfest** [comment]
Thanks @conneryn! 

1) No problem on my side. Workaround good to go.
2) I had the same issue. While the workaround does it job I miss the descriptions. I'm a huge fan of `kubectl explain` and we'd be forcing the user to go to the spec source to understand each field. Can we use `kubectl replace --force -f -` instead? 

Gonna review the files to check If I missed something but it's looking great so far :D

File: main.go [hunk 1]
```diff
@@ -51,6 +51,7 @@ func init() {
 
 	utilruntime.Must(torv1alpha1.AddToScheme(scheme))
 	utilruntime.Must(torv1alpha2.AddToScheme(scheme))
+	utilruntime.Must(torv1alpha2.AddToScheme(scheme))
```
Comment (bugfest) [inline main.go:54]: duplicated

**Review by bugfest — CHANGES_REQUESTED** [review]
just a couple of minor things

**Review by bugfest — APPROVED** [review]

**Pull request merged**